Introduce compute listener

dnhatn opened the pull request:
Currently, if a child request fails, we automatically trigger cancellation for ES|QL requests. This can result in TaskCancelledException being collected by the RefCountingListener first, which then returns that exception to the caller. For example, if we encounter a CircuitBreakingException (429), we might incorrectly return a TaskCancelledException (400) instead.

This change introduces the ComputeListener, a variant of RefCountingListener, which ignores unimportant exceptions and selects the most appropriate exception to return to the caller. I also integrated the following features into ComputeListener to simplify ComputeService:

1. Automatic cancellation of sub-tasks on failure.
2. Collection of profiles from sub-tasks.
3. Collection of response headers from sub-tasks.

dnhatn committed:
Introduce compute listener

elasticsearchmachine commented:
Hi @dnhatn, I've created a changelog YAML for you.

dnhatn committed:
Update docs/changelog/110400.yaml

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/FailureCollector.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Collects exceptions occurred in compute, specifically ignores trivial exceptions such as {@link TaskCancelledException},
```
The comment isn't quite right I think - if there aren't any exceptions captured it'll capture the first "trivial exception". And it'll replace that with the first non-trivial exception it sees.

nik9000 commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/operator/FailureCollector.java:
```diff
@@ -0,0 +1,54 @@
+            if (ExceptionsHelper.unwrapCause(first) != ExceptionsHelper.unwrapCause(e)) {
+                if (exceptionPermits.tryAcquire()) {
+                    first.addSuppressed(e);
+                }
```
`getAndUpdate` is documented as requiring a side-effect-free function because it'll be retried sometimes. I *think* this is safe, but it's not really side effect free. Maybe it'd be more readable if we used an atomicarray or something. Either way, I spent a while reading this code and thinking about it, but I don't think the performance is all that critical here because it's deep into exception path.

nik9000 reviewed (commented):
:+1: on collecting the listener behavior into a named class.

Do you think it's appropriate to pull that class into it's own top level file? Do you think it's appropriate to make a test case specifically for it?

I know this has been quite the distraction and I'm proposing more distractions, but the listener behavior is pretty complex and a test might help other people to hold it in their head.

dnhatn committed:
Add tests

dnhatn committed:
Merge remote-tracking branch 'elastic/main' into compute-listener

dnhatn commented:
@nik9000 Thank you for the feedback. I have made the ComputeLister its own class and added tests for it. I also updated the documentation and the implementation of FailureCollector to use concurrent queues. This should be ready for review again. Could you please take another look?

dnhatn committed:
stylecheck

dnhatn commented:
Thanks Nik!

elasticsearchmachine commented:
## 💔 Backport failed
| Status | Branch | Result |
|:------:|:------:|:------:|
| ❌ |  8.14  | Commit could not be cherrypicked due to conflicts |
| ❌ |  8.15  | Commit could not be cherrypicked due to conflicts |

You can use [sqren/backport](https://github.com/sqren/backport) to manually backport by running `backport --upstream elastic/elasticsearch --pr 110400`
